### ISSUE TITLE
Is there a way to schedule a recurring task on the fly?

### ISSUE THREAD

**Issue by vivekkothari** [kagkarlsson/db-scheduler#246]
I am a simple task which puts a SQS message.
The data it takes is also very simple, a sqsUrl and message body.
I have certain case where some of the instances of the job need to run every day.
For the rest (which are one-off instances) i am able to use db-schedule easily.

**Comment by kagkarlsson** [comment]
Hi! What was the actual question here?

Sounds a bit like you need two different Tasks, one with a schedule (recurring) and one for onetime-tasks

**Comment by vivekkothari** [comment]
```
var dailyCronEvent = recurring("daily-cron", Schedules.daily(UTC, LocalTime.MIDNIGHT))
        .execute((inst, ctx) -> executeJob());
```
This cron has to be coded in and specified in the `startTasks` method when starting the scheduler.
Quartz provides a way to schedule cron on the fly. Can we do something similar?

**Comment by kibertoad** [comment]
We need this feature as well. Happy to work on a PR for this if accepted. It seems rather strange that only one-off tasks can be scheduled dynamically after scheduler is already running.

**Comment by kibertoad** [comment]
@kagkarlsson API proposal for the feature:
```
    public <T extends Task<?> & OnStartup> void scheduleRecurring(T task) {
        // should execute those task as defined in their description
    }
```

**Comment by kagkarlsson** [comment]
It is currently possible, but the api is a bit unintuitive atm. See https://github.com/kagkarlsson/db-scheduler/blob/master/examples/features/src/main/java/com/github/kagkarlsson/examples/PersistentDynamicScheduleMain.java#L33

@kibertoad agreed, something along those lines would be nice. I probably need to draft some code to see what will work best and most generally. I think the parameter need to be something that contains task-name, instance-id, optionally initial data, optionally initial execution-time (default to now)

**Comment by rpsandiford** [comment]
We're evaluating this software, and this capability (i.e. ability to add a new recurring task to an existing scheduler, or remove a recurring task).

Our use case is that we have a DB with 'customer' records - and for each customer record, we will have a monitor type process that runs periodically.  Customer records can be added (in which case we'd want to add a monitor process) or deleted / disabled (in which case we'd want to remove the corresponding monitor process).

I would also likely have bandwidth to help out with a PR for this...  (Even if it's just to help test it :)

thanks!

**Comment by kagkarlsson** [comment]
@rpsandiford 
I need to clean up some of the issues. Some of them hint at limitations of db-scheduler that is not there..

Out of the box, `RecurringTask` is limited to a single instance, but there is nothing preventing you from creating a custom recurring task that supports multiple instances better. The only thing is that you have to supply the first execution-time to the scheduler when scheduling the instances. I am working on an improvement there, but that is more a cosmetic change than functional one.

Would the schedule be the same for all instances?

I much appreciate the offer for help. Maybe a PR for an example for your use-case? (here: https://github.com/kagkarlsson/db-scheduler/tree/master/examples/features/src/main/java/com/github/kagkarlsson/examples)

**Comment by kagkarlsson** [comment]
If you don't see how to accomplish this, let me know..

**Comment by kagkarlsson** [comment]
Btw, do you want to be able to change the Schedule by just deploying a new version of the app like "normal" RecurringTasks?

**Comment by rpsandiford** [comment]
Hi, Gustav.
I'll do some more playing with the code taking into account your suggestion, but right now it looks like:
* When creating the scheduler, you can supply as many Tasks as you want - whether recurring or not (though I haven't actually tried multiple recurring tasks).
* Once the scheduler is up and running, you can add more TaskInstance type tasks.

The schedule would not be the same for each recurring task (one of the configurable elements for a customer is how often to run the monitor process for that customer.)  Another configurable element is whether or not the customer is 'active' (allows us to suspend operations if the customer has downtime, or as we're initially setting up the configuration but not yet ready to actually start communications)

Those requirements are driving the desire to be able to 
* Add a new RecurringTask to an existing scheduler - we want to be able to share the same thread pool, for example.  I know I could create a separate scheduler for each customer - but then our thread count, even at one thread per scheduler, would go through the roof.
* Remove an existing Recurring Task from an existing scheduler (or suspend it or some other way to stop it executing on the schedule).

Since these need to be responsive to changes made in the customer data, stopping and starting the app (or deploying a new version of the app) would be really awkward.  We'd want the new Recurring Task (or the suspension / removal of an existing RecurringTask) to be reflected in all nodes of the app.

I'll give some thought to how to set up an example that would illustrate more explicitly our situation.

Thanks!

**Comment by rpsandiford** [comment]
Actually - it just dawned on me (I'm a bit slow, apparently) - just adding a new Task on one node of a multi-node system doesn't automatically mean the other nodes 'know' about that task.  The persistence table is only for tracking and coordinating tasks that each node knows about.  i.e. just putting a new row into that table won't automagically mean all nodes 'know' about that new entry and what to do with it.  When adding a new Task (or removing one) - each node will need to coordinate to create a local instance of that task so that when it is triggered from the table, whichever node picks it up can invoke the corresponding task on that node.

So - I'll think a bit more about the possibilities here, and what would be needed to accomplish what I'd like to see happen...

**Comment by kagkarlsson** [comment]
@rpsandiford I think I missed your last two comments here, sorry!

I the meantime I have thought about some improvents that might make things easier for you. It is a work in progress, but feedback is appreciated. See PR #257

**Comment by kagkarlsson** [comment]
Improved api for recurring-tasks that can be scheduled on the fly released in v11.0.
Have a look at the example at [RecurringTaskWithPersistentScheduleMain.java](https://github.com/kagkarlsson/db-scheduler/blob/master/examples/features/src/main/java/com/github/kagkarlsson/examples/RecurringTaskWithPersistentScheduleMain.java).
Closing this issue.

**Comment by kagkarlsson** [comment]
🎉 This issue has been resolved in `v16.9.0` ([Release Notes](https://github.com/kagkarlsson/db-scheduler/releases/tag/v16.9.0))
